### PR TITLE
Add REST API endpoints required to move client to REST. And additional refactoring

### PR DESCRIPTION
**Pull request by TShapinsky**
Changes:

- Unit tests are now run with mongo/redis/etc... (Open to comments, but I feel like this is the easiest way rather than continuing to create more and more complex mock classes)
- Added the following REST API endpoints
    - /sites/:id/time
    - /sites/:id/points
    - /sites/:id/points/inputs
    - /sites/:id/points/outputs
    - /sites/:id/points/:id
    - /models
    - /models/upload
    - /models/:id/createRun
- Created API method in web to send messages to runs
- Model objects are now created when you request an upload form rather than when a run uses it
- Run and Point classes now only exist in `models.py`
- Point objects now directly access redis via their `value` property (no need to call redis directly)
- `AlfalfaConnectionsManager` added as a singleton too provide connections to classes while reducing overall instance counts (this could possible be moved to a module if we wanted?)
- Enums consolidated
- Tests updated to work with next version of the alfalfa-client

TODO:

- [x] `/sites` web ui appears to be a little broken
- [ ] Add documentation to the REST api for new endpoints

### CODE REVIEW

File: alfalfa_worker/lib/models.py [hunk 1]
```diff
@@ -396,10 +405,62 @@ class Rec(TimestampedDocument):
...
+
+
+class Point(TimestampedDocument):
+    meta = {
+        'collection': 'point',
+        'indexes': ['ref_id', 'run'],
+    }
+
+    def __init__(self, *args, **values):
+        connections_manager = AlafalfaConnectionsManager()
```
Comment (TShapinsky) [inline alfalfa_worker/lib/models.py:430]: @nllong

File: alfalfa_web/server/api-v2.js [hunk 2]
```diff
@@ -304,6 +353,61 @@ router.get("/version", (req, res) => {
...
+    .then((data) => res.json({ data }))
+    .catch(() => res.sendStatus(500));
+});
+
+router.post("/models/:id/createRun", (req, res) => {
+  const { id: modelID } = req.params;
+  api
+    .createRunFromModel(modelID)
+    .then((runID) => {
+      res.json({ runID: runID });
```
Comment (axelstudios) [inline alfalfa_web/server/api-v2.js:368]: Simplify to `{ runID }`

File: alfalfa_web/server/api-v2.js [hunk 3]
```diff
@@ -304,6 +353,61 @@ router.get("/version", (req, res) => {
...
+    .catch((err) => {
+      console.log(err);
+      res.sendStatus(500);
+    });
+});
+
+// Create a post url for file uploads
+// from a browser
+router.post("/models/upload", (req, res) => {
+  const { modelName: modelName } = req.body;
```
Comment (axelstudios) [inline alfalfa_web/server/api-v2.js:379]: Simplify to `{ modelName }`

File: alfalfa_web/server/api-v2.js [hunk 4]
```diff
@@ -304,6 +353,61 @@ router.get("/version", (req, res) => {
...
+      // if you're running locally and using internal Docker networking ( "http://minio:9000")
+      // as your S3_URL, you need to specify an alternate S3_URL_EXTERNAL to POST to, ie "http://localhost:9000"
+      if (process.env.S3_URL_EXTERNAL) {
+        data.url = `${process.env.S3_URL_EXTERNAL}/${process.env.S3_BUCKET}`;
+      } else {
+        data.url = `${process.env.S3_URL}/${process.env.S3_BUCKET}`;
+      }
+      data.modelID = modelID;
+      res.send(JSON.stringify(data));
+      res.end();
```
Comment (axelstudios) [inline alfalfa_web/server/api-v2.js:406]: Can we simplify this to to `res.json(data)` or do we explicitly want a text/plain response?

File: alfalfa_web/server/api.js [hunk 5]
```diff
@@ -1,10 +1,12 @@
 import AWS from "aws-sdk";
 import { v1 as uuidv1 } from "uuid";
 import { Advancer } from "./advancer";
-import { del, mapHaystack, reduceById, scan } from "./utils";
+import { writePoint } from "./dbops";
+import { del, getHashValue, getPointKey, mapHaystack, reduceById, scan, setHashValue, getHash } from "./utils";
 
 class AlfalfaAPI {
   constructor(db, redis) {
+    this.db = db;
```
Comment (axelstudios) [inline alfalfa_web/server/api.js:9]: We can remove the `this.advancer = ...` below now, right?  And the import?

File: alfalfa_web/server/api.js [hunk 6]
```diff
@@ -29,12 +32,14 @@ class AlfalfaAPI {
       const sites = (await this.sites.find().toArray()).map(mapHaystack).reduce(reduceById, {});
 
       for await (const run of this.runs.find()) {
+        let model = models[run.model];
```
Comment (axelstudios) [inline alfalfa_web/server/api.js:35]: const

File: alfalfa_web/server/api.js [hunk 7]
```diff
@@ -65,6 +71,63 @@ class AlfalfaAPI {
...
+        if (pointType) {
+          query.point_type = pointType;
+        }
+        for await (const point of this.points.find(query)) {
+          const pointDict = {
+            id: point.ref_id,
+            name: point.name,
+            type: point.point_type
+          };
+          if (getValue && point.point_type != "INPUT") {
```
Comment (axelstudios) [inline alfalfa_web/server/api.js:97]: `!==`

File: pyproject.toml [hunk 8]
```diff
@@ -55,7 +55,7 @@ influxdb = "^5.3.1"
 # required for tests to pass right now. There are often issues when pinning to a branch
 # and calling poetry update. I have had to completely remove my venv and reinstall in order
 # to get the latest updates.
-alfalfa-client = "^0.4.0"
+alfalfa-client = {git = "https://git@github.com/nrel/alfalfa-client.git", branch = "rest_api"}
```
Comment (axelstudios) [inline pyproject.toml:58]: We'll be changing this after merging, right?

Comment (TShapinsky) [reply pyproject.toml:58]: Yeah, the alfalfa client change to rest API hasn't been merged or released on that side yet

**Review by axelstudios — APPROVED** [review]

**Pull request merged**